appinventor/blocklyeditor/src/component_database.js

```diff
@@ -504,6 +514,45 @@ Blockly.ComponentDatabase.prototype.processOption = function(option) {
   };
 }
 
+/**
+ * Processes data defining an asset filter (from simple_components.json) and
+ * returns a HelperKey pointing to the filter.
+ * @param {!Object} data The data defining the filter.
+ * @return {!HelperKey} The key associated with the filter.
+ */
+Blockly.ComponentDatabase.prototype.processAssetHelper = function(data) {
+  var filter = data.filter;
```

Comment:
There is a minor optimization that can be done here if the filter is sorted where you use binary search on the registered filters using the target filter. Since the elements are ordered, if they don't match starting from the first element you can fail quickly and move to the next potential registered element. I'm not sure it would warrant a rewrite this late in the game but you may just want to note it as a TODO.

Reply:
I think .every does fail quickly, so now that we are sorting the filter elements inside the ComponentProcessor we should get that benefit :D

But I think that you're advising that we should sort the filters in this.filters_ in addition to the elements in each filter. I'll add a TODO for that.

Reply:
You should change the return type of `Blockly.Blocks.Utilities.assetKeyToBlocklyType` to match this function then as it is the only place that it is called from. This return allows for undefined but the caller does not.